File: TwentyFortyEight/TFE.m

#import "TFE.h"
#import "TFENode.h"
#import "NSArray+TFEFunctional.h"
#import "NSIndexSet+TFEHelpers.h"

NSString * const kTFENodeKey = @"TFENode";
NSString * const kTFEMoveKey = @"TFEMove";
NSString * const kTFEMoveIsSpawnKey = @"TFEMoveIsSpawn";
NSString * const kTFEMoveIsComboKey = @"TFEMoveIsCombo";

#pragma mark - Private function decls

/** Returns the indexes for the grid line furthest in the given direction; 
 *  these are the squares where spawning is not allowed after a move has 
 *  been made.
 */
NSIndexSet * TFEDisallowedSquaresByDirection(TFENodeDirection d);
/** Returns an NSMutableArray of 16 NSNull. */
NSMutableArray * TFENullArray(void);
/** Returns the indexes in squares that do not contain NSNull. */
NSIndexSet * TFEIndexesOfUnoccupiedSquares(NSArray * squares);
/** Given a list of nodes, attempts to "slide" them towards the front of the
 *  list, combining adjacent equal-valued nodes, and moving nodes through
 *  empty spaces.
 *  Removes nulls before sliding.
 *  Returns the reconfigured row, or nil if nothing actually moved.
 */
NSArray * TFESlideRow(NSArray * row);

/** Does the actual work for the public function 
 *  TFESpawnNewNodeExcludingDirection(), which will pass in the result of
 *  TFEDisallowedSquaresByDirection()
 */
NSArray * TFESpawnNewNode(NSArray * grid,
                       NSIndexSet * disallowedIndexes,
                       NSDictionary ** spawn);

/** Returns a dictionary describing a node's movement: the node,
 *  the destination square, whether the node is being combined with another,
 *  or whether the move is a spawn.
 */
NSDictionary * TFEMoveDescription(TFENode * node, NSUInteger destination,
                               BOOL isCombo, BOOL isSpawn);

#pragma mark - Public fuctions

NSArray * TFEBuildGrid(NSArray **spawns)
{
    NSArray * grid = (NSArray *)TFENullArray();
    
    NSDictionary * firstSpawn;
    NSDictionary * secondSpawn;
    grid = TFESpawnNewNode(grid, nil, &firstSpawn);
    grid = TFESpawnNewNode(grid, nil, &secondSpawn);
    
    *spawns = @[firstSpawn, secondSpawn];
    
    return grid;
}

static const NSUInteger gridLinesByDirection[4][4][4] =
    {{{0, 1, 2, 3}, {4, 5, 6, 7}, {8, 9, 10, 11}, {12, 13, 14, 15}},
     {{12, 8, 4, 0}, {13, 9, 5, 1}, {14, 10, 6, 2}, {15, 11, 7, 3}},
     {{3, 2, 1, 0}, {7, 6, 5, 4}, {11, 10, 9, 8}, {15, 14, 13, 12}},
     {{0, 4, 8, 12}, {1, 5, 9, 13}, {2, 6, 10, 14}, {3, 7, 11, 15}}};

NSArray * TFEMoveNodesInDirection(NSArray * grid,
                               TFENodeDirection direction,
                               NSArray ** moves)
{
    NSMutableArray * newGrid = TFENullArray();
    NSMutableArray * newMoves = nil;
    
    for( int row_i = 0; row_i < 4; row_i++ ){
        
        // Grab nodes for row and clear out square master list
        // Can't do this with an index set because they need to be in reverse
        // order when moving in two of the four directions.
        NSMutableArray * rowNodes = [NSMutableArray array];
        for( int i = 0; i < 4; i++ ){
            NSUInteger idx = gridLinesByDirection[direction][row_i][i];
            rowNodes[i] = grid[idx];
        }
        
        NSArray * slidRow = TFESlideRow(rowNodes);
        // No movement, just copy over nodes and move to the next row.
        if( !slidRow ){
            for( int i = 0; i < 4; i++ ){
                NSUInteger idx = gridLinesByDirection[direction][row_i][i];
                newGrid[idx] = grid[idx];
            }
            continue;
        }
        
        newMoves = newMoves ?: [NSMutableArray array];
        [slidRow enumerateObjectsUsingBlock:
             ^(id element, NSUInteger idx, BOOL * stop)
             {
                 NSUInteger destSquare =
                                  gridLinesByDirection[direction][row_i][idx];
                 if( [element isKindOfClass:[NSArray class]] ){
                     for( TFENode * node in element ){
                         [newMoves addObject:
                                TFEMoveDescription(node, destSquare, YES, NO)];
                     }
                     uint32_t newVal = [(TFENode*)element[0] value] * 2;
                     TFENode * combinedNode = [TFENode nodeWithValue:newVal];
                     newGrid[destSquare] = combinedNode;
                     [newMoves addObject:
                        TFEMoveDescription(combinedNode, destSquare, NO, YES)];
                 }
                 else {
                     newGrid[destSquare] = element;
                     [newMoves addObject:
                        TFEMoveDescription(element, destSquare, NO, NO)];
                 }
             }];
    }
    
    // Still nil if no moves were made
    *moves = newMoves;
    return newGrid;
}

NSArray * TFESpawnNewNodeExcludingDirection(NSArray * grid,
                                         TFENodeDirection direction,
                                         NSDictionary ** spawn)
{
    return TFESpawnNewNode(grid, TFEDisallowedSquaresByDirection(direction), spawn);
}



BOOL TFEIsAWinner(NSArray * grid)
{
    static TFENode * twentyFortyEight;
    if( !twentyFortyEight ){
        twentyFortyEight = [TFENode nodeWithValue:2048];
    }
    
    return [grid containsObject:twentyFortyEight];
}

BOOL TFEIsALoser(NSArray * grid)
{
    // If the board is not full, can't have lost yet
    if( [TFEIndexesOfUnoccupiedSquares(grid) count] > 0 ){
        return NO;
    }
    
    // Check movement in all directions.
    NSArray * moves = nil;
    for( TFENodeDirection d = 0; d < 4; d++ ){
        id __unused _ = TFEMoveNodesInDirection(grid, d, &moves);
        // If movement is possible in _any_ direction, haven't lost.
        if( moves ){
            break;
        }
    }
    
    return (moves == nil);
}

#pragma mark - Private functions

NSIndexSet * TFEDisallowedSquaresByDirection(TFENodeDirection d)
{
    static const NSUInteger left[4] = {12, 8, 4, 0};
    static NSIndexSet * leftSet;
    if( !leftSet ){
        leftSet = [NSIndexSet TFEIndexSetWithIntegers:left count:4];
    }
    static const NSUInteger up[4] = {12, 13, 14, 15};
    static NSIndexSet * upSet;
    if( !upSet ){
        upSet = [NSIndexSet TFEIndexSetWithIntegers:up count:4];
    }
    static const NSUInteger right[4] = {15, 11, 7, 3};
    static NSIndexSet * rightSet;
    if( !rightSet ){
        rightSet = [NSIndexSet TFEIndexSetWithIntegers:right count:4];
    }
    static const NSUInteger down[4] = {0, 1, 2, 3};
    static NSIndexSet * downSet;
    if( !downSet ){
        downSet = [NSIndexSet TFEIndexSetWithIntegers:down count:4];
    }
    
    static NSArray * list;
    if( !list ){
        list = @[leftSet, upSet, rightSet, downSet];
    }
    
    return list[d];
}

NSMutableArray * TFENullArray(void)
{
    NSMutableArray * grid = [NSMutableArray array];
    for( int i = 0; i < 16; i++ ){
        grid[i] = [NSNull null];
    }
    
    return grid;
}

NSIndexSet * TFEIndexesOfUnoccupiedSquares(NSArray * grid)
{
    return [grid indexesOfObjectsPassingTest:
                ^BOOL (id obj, NSUInteger idx, BOOL * stop)
                  {
                      return obj == [NSNull null];
                  }];
}

NSArray * TFESlideRow(NSArray * row)
{
    NSArray * realNodes = [row filteredArrayUsingPredicate:
                            [NSPredicate predicateWithFormat:@"self != nil"]];
    if( [realNodes count] == 0 ){
        return nil;
    }
    
    TFEFolder slider =
        ^NSMutableArray * (NSMutableArray * accum, TFENode * node)
        {
            id newVal = node;
            id lastVal = [accum lastObject];
            // lastVal is either NSArray of two nodes or a node
            // If it's a node, and equal, combine them into an NSArray
            // for further processing
            if( [lastVal isEqual:node] ){
                 [accum removeLastObject];
                 newVal = @[lastVal, node];
             }

             // Otherwise just add the new node to the result
             [accum addObject:newVal];
             return accum;
        };
    
    NSMutableArray * initial = [NSMutableArray arrayWithObject:realNodes[0]];
    NSArray * slidRow = [[realNodes TFETail] TFEFoldWithInitial:initial
                                                    usingFolder:slider];
    // slidRow must have at least one object in it.
    // Adding three nulls and then cutting off at four is easier than
    // counting and constructing a new mutable array again.
    NSArray * reconstructedRow =
        [[slidRow arrayByAddingObjectsFromArray:
                                @[[NSNull null], [NSNull null], [NSNull null]]]
            subarrayWithRange:(NSRange){0,4}];
    BOOL didMove = ![row isEqualToArray:reconstructedRow];
    return didMove ? slidRow : nil;
}

NSArray * TFESpawnNewNode(NSArray * grid,
                       NSIndexSet * disallowedIndexes,
                       NSDictionary ** spawn)
{
    NSMutableIndexSet * unoccupiedSquares =
                    [TFEIndexesOfUnoccupiedSquares(grid) mutableCopy];
    if( disallowedIndexes ){
        [unoccupiedSquares removeIndexes:disallowedIndexes];
    }
    
    NSUInteger idx = [unoccupiedSquares TFERandomIndex];
    
    // Bias heavily towards new value being a 2
    uint32_t draw = arc4random_uniform(10);
    uint32_t val = (draw < 9) ? 2 : 4;
    TFENode * node = [TFENode nodeWithValue:val];
    
    NSMutableArray * newGrid = [grid mutableCopy];
    [newGrid replaceObjectAtIndex:idx
                       withObject:node];
    
    *spawn = TFEMoveDescription(node, idx, NO, YES);
    
    return newGrid;
}
         
NSDictionary * TFEMoveDescription(TFENode * node, NSUInteger destination,
                                  BOOL isCombo, BOOL isSpawn)
{
    return @{kTFENodeKey : node,
             kTFEMoveKey : @(destination),
             kTFEMoveIsComboKey : @(isCombo),
             kTFEMoveIsSpawnKey : @(isSpawn)};
}

#pragma mark - Debug functions

char * center_string(const char * s, int width);
void drawGrid(NSArray * grid);
NSMutableArray * fakeGrid(uint32_t v, ...);

char * center_string(const char * s, int width)
{
    int len = (int)strlen(s);
    int pad_half_len = (width - len) / 2;
    char * r = calloc(7, sizeof(char));
    for( int i = 0; i < pad_half_len; i++ ){
        r = strcat(r, " ");
    }
    r = strcat(r, s);
    for( int i = 0; i < pad_half_len; i++ ){
        r = strcat(r, " ");
    }
    
    return r;
}

void drawGrid(NSArray * grid)
{
    for( int row = 0; row < 4; row++ ){
        for( int col = 0; col < 4; col++ ){
            NSUInteger idx = (row * 4) + col;
            id elem = grid[idx];
            NSString * desc;
            if( elem == [NSNull null] ){
                desc = @" __ ";
            }
            else {
                desc = [NSString stringWithFormat:@" %2d ",
                                                  [(TFENode*)elem value]];
            }
            printf("%s", center_string([desc UTF8String], 6));
        }
        printf("\n");
    }
    printf("\n");
}

NSMutableArray * fakeGrid(uint32_t v, ...)
{
    NSMutableArray * grid = [NSMutableArray array];
    va_list args;
    va_start(args, v);
    uint32_t value = v;
    while( value != UINT32_MAX ){
        id new = [NSNull null];
        if( value != 0 ){
            new = [TFENode nodeWithValue:value];
        }
        [grid addObject:new];
        value = va_arg(args, uint32_t);
    }
    va_end(args);
    
    return grid;
}

File: Common/Game Core/TFE.swift
//
//  TFE.swift
//  TwentyFortyEight
//
//  Created by Joshua Caswell on 10/4/16.
//  Copyright © 2016 Josh Caswell. All rights reserved.
//

import Foundation

/**
 * Construct a starting grid; the grid is empty except for two tiles created
 * by spawnNewTile(on:)
 *
 * Returns an array of `TileMove`s describing the spawns, and the new grid.
 */
public func buildGrid<T : Tile>() -> ([TileMove<T>], [T?])
{
    var grid: [T?] = nullGrid()
    
    let firstSpawn: TileMove<T>
    let secondSpawn: TileMove<T>
    
    (firstSpawn, grid) = spawnNewTile(on: grid)
    (secondSpawn, grid) = spawnNewTile(on: grid)
    
    return ([firstSpawn, secondSpawn], grid)
}

/**
 * Attempt to slide all the tiles towards the given direction.
 *
 * Returns an array of `TileMove`s, or `nil` if nothing moved, and the grid as
 * reconfigured after all movement.
 */
public func moveTiles<T : Tile>(_ grid: [T?], inDirection direction: SlideDirection) -> ([TileMove<T>]?, [T?])
{
    var newGrid: [T?] = nullGrid()
    var moves: [TileMove<T>]? = nil
    
    // The by-row and intra-row traversal orders depend on the direction
    // of movement. GridIndexes holds the square positions in the
    // correct orders
    for rowIndexes in GridIndexes.by(direction) {
        
        let row = grid.elements(at: rowIndexes)
        
        let slid = slideRow(row)
        
        // No movement, just copy tiles over and continue with the next row.
        guard let slidRow = slid else {
            
            for index in rowIndexes {
                newGrid[index] = grid[index]
            }
            
            continue
        }
        
        moves = moves ?? []
        
        let newMoves: [TileMove<T>]
        (newMoves, newGrid) = processMoves(forSlidRow: slidRow, rowIndexes: rowIndexes, onGrid: newGrid)
        moves?.append(contentsOf: newMoves)
    }
    
    return (moves, newGrid)
}


/**
 * Create a new tile at a random index selected from those that are both
 * unoccupied and not in the line furthest in the disallowed direction.
 * The tile has value 2 or 4, chosen at random.
 *
 * Returns a `TileMove` object describing the spawn and the grid with the
 * new tile added.
 */
public func spawnNewTile<T : Tile>(on grid: [T?], excluding direction: SlideDirection) -> (TileMove<T>, [T?])
{
    let disallowedIndexes = TrailingSquares.awayFrom(direction)
    return spawnNewTile(on: grid, excluding: disallowedIndexes)
}

public func isAWinner<T : Tile>(_ grid: [T?]) -> Bool
{
    return grid.contains(where: { $0?.value == 2048 })
}

public func isALoser<T : Tile>(_ grid: [T?]) -> Bool
{
    // If the board isn't full, we can't have lost yet
    guard !indexesOfUnoccupiedSquares(grid).isEmpty else {
        return false
    }
    
    // Try moving in all directions
    let directions: [SlideDirection] = [.left, .up, .right, .down]
    for direction in directions {
        
        let (moves, _) = moveTiles(grid, inDirection: direction)
        
        // If movement is possible in _any_ direction, we haven't lost
        guard moves == nil else {
            return false
        }
    }
    
    return true
}

public func score<T : Tile>(forMoves moves: [TileMove<T>]) -> Int
{
    return moves.reduce(0, { (score, move) in
        
        guard move.isCombination else {
            return 0
        }
        
        return score + move.tile.value
    })
}

//MARK: - Internal functions

/** Array of `Optional<T>.None` representing an empty grid. */
func nullGrid<T : Tile>() -> [T?]
{
    return Array(repeating: nil, count: 16)
}

/**
 * Given a list of optional `Tile`s, representing the game grid, return a set
 * with the indexes of all nil elements.
 */
func indexesOfUnoccupiedSquares<T : Tile>(_ grid: [T?]) -> Set<Int>
{
    var set: Set<Int> = []
    
    for (i, v) in grid.enumerated() {
        if v == nil {
            set.insert(i)
        }
    }
    
    return set
}

/**
 * Given a list of `SlidTile`s and the grid indexes of the corresponding row,
 * deconstruct the enums into `TileMove`s and update the tile locations in the
 * grid, or create new tiles for combinations, as necessary.
 *
 * Returns the list of moves and the updated grid.
 */
func processMoves<T : Tile>(forSlidRow row: [SlidTile<T>], rowIndexes indexes: [Int], onGrid grid: [T?]) -> ([TileMove<T>], [T?])
{
    var newGrid = grid
    var moves: [TileMove<T>] = []
    
    for (destination, slidTile) in zip(indexes, row) {
    
        let tileForDestination: T
    
        switch slidTile {
        
            case .empty:
                continue
            case let .solo(tile):
                tileForDestination = tile
                moves.append(TileMove(tile: tile, destination: destination))
            case let .combined(firstTile, secondTile):
                moves.append(TileMove(tile: firstTile, destination: destination, isCombination: true))
                moves.append(TileMove(tile: secondTile, destination: destination, isCombination: true))
                let comboTile = T(value: firstTile.value * 2)
                tileForDestination = comboTile
                moves.append(TileMove(tile: comboTile, destination: destination, isSpawn: true))
        }
    
        newGrid[destination] = tileForDestination
    }
    
    return (moves, newGrid)
}

/**
 * Given an array of optional `Tile`s representing a grid row, move the values
 * leftwards (towards the beginning of the array), allowing tiles that contain
 * values to skip over empty squares and combining adjacent tiles that have the
 * same value.
 *
 * Return `nil` if nothing changed, or an array of `SlidTile`s representing the
 * moved tiles for further processing.
 */
func slideRow<T : Tile>(_ row: [T?]) -> [SlidTile<T>]?
{
    let realTiles: [T] = row.flatMap({ $0 })
    
    guard !realTiles.isEmpty else {
        return nil
    }
    
    /**
     * Perform the combinations of adjacent tiles as a left fold:
     * `accum` initially holds the first element of the row as a `SlidTile`,
     * then collects each new element as it is processed. When two `.Solo` tiles
     * are moved into one another, they create a single `.Combined`, with the
     * same associated value.
     */
    func slide(accum: [SlidTile<T>], next: T) -> [SlidTile<T>]
    {
        switch accum.last! {
            
            case let .solo(last) where last == next:
                return accum.dropLast() + [SlidTile(last, next)]
            default:
                return accum + [SlidTile(next)]
        }
    }
    
    let (first, rest) = (realTiles.first!, realTiles.dropFirst())
    let slidRow = rest.reduce([SlidTile(first)], slide)
    
    // "Fill out" the remainer of the row with SlidTile.Empty values
    let emptiesCount = 4 - slidRow.count
    let filledOutRow = slidRow + Array(repeating: SlidTile(), count: emptiesCount)
    
    // If the slid row after adding empties is equivalent to the original row mapped to
    // SlidTiles, nothing changed.
    let wrappedOriginalRow = row.map({ SlidTile<T>($0) })
    let didMove = (filledOutRow != wrappedOriginalRow)
    
    return didMove ? slidRow : nil
}

/**
 * Does the actual work for the public function `TFESpawnNewTile(on:excluding:)`,
 * which will pass in the disallowed directions.
 *
 * Given the grid and a set of positions that are invalid, add a square with
 * value either 2 or 4 to any empty and valid square, then return the new grid.
 */
func spawnNewTile<T : Tile>(on grid: [T?], excluding disallowedIndexes: Set<Int> = []) -> (TileMove<T>, [T?])
{
    let unoccupiedSquares = indexesOfUnoccupiedSquares(grid)
    let allowedLocations = unoccupiedSquares.subtracting(disallowedIndexes)
    let spawnLocation = allowedLocations.randomInt()!
    
    var newGrid = grid
    // Bias heavily towards new value being a 2
    let pick = arc4random_uniform(10)
    let newValue = (pick < 9) ? 2 : 4
    let tile = T(value: newValue)
    
    let spawn = TileMove(tile: tile, destination: spawnLocation, isSpawn: true)
    newGrid[spawnLocation] = tile
    
    return (spawn, newGrid)
}
